### ISSUE TITLE
group-by, aggregate, pivot via xforms?

### ISSUE THREAD

**Issue by aaelony** [cgrand/xforms#2]
HI Christophe, this library looks really great!  

I am not yet groking it fully though.  I'd like to group-by 2 keys, then aggregate 2 other keys, then pivot the data. 

 e.g. To start from a collection of Maps like 

```

(def dat [{:cohort-month "2015-01" :followup-month "2015-02" :x1 100 :x2 200 :strata "class 1"}
          {:cohort-month "2015-01" :followup-month "2015-02" :x1 5 :x2 7 :strata "class 1"}
          {:cohort-month "2015-01" :followup-month "2015-03" :x1 103 :x2 196 :strata "class 2"}
          {:cohort-month "2015-02" :followup-month "2015-03" :x1 110 :x2 201 :strata "class 2"}
          {:cohort-month "2015-01" :followup-month "2015-04" :x1 107 :x2 193 :strata "class 3" }
          ])
```

and then use transducers that aggregate the `:x1` and `:x2` keys by `:cohort-month` and `:followup-up month`, and then pivot it finally into something like this:

```
(def dat-pivot-desired [[nil :followup-months nil nil]
                        [:cohort-month "2015-02" "2015-03" "2015-04" ]
                        ["2015-01" {:x1 105 :x2 207} {:x1 103 :x2 196} {:x1 107 :x2 193}]
                        ["2015-02" {} {:x1 110 :x2 201} {}]
                        ])
```

Ideally, there could be any number of :x1 ... :xN keys ??

Thank-you,
Avram

**Comment by cgrand** [comment]
Hi Avram!

Nice to hear back from you! The nested aggregation is not a problem.
However I'll have to think about the best way to perform the pivot.

I'm surprised by the desired output (with header lines). It's for display
in a grid widget?

Le lundi 14 septembre 2015, aaelony notifications@github.com a écrit :

> HI Christophe, this library looks really great!
> 
> I am not yet groking it fully though. I'd like to group-by 2 keys, then
> aggregate 2 other keys, then pivot the data.
> 
> e.g. To start from a collection of Maps like
> 
> (def dat [{:cohort-month "2015-01" :followup-month "2015-02" :x1 100 :x2 200 :strata "class 1"}
>           {:cohort-month "2015-01" :followup-month "2015-02" :x1 5 :x2 7 :strata "class 1"}
>           {:cohort-month "2015-01" :followup-month "2015-03" :x1 103 :x2 196 :strata "class 2"}
>           {:cohort-month "2015-02" :followup-month "2015-03" :x1 110 :x2 201 :strata "class 2"}
>           {:cohort-month "2015-01" :followup-month "2015-04" :x1 107 :x2 193 :strata "class 3" }
>           ])
> 
> and then use transducers that aggregate the :x1 and :x2 keys by
> :cohort-month and :followup-up month, and then pivot it finally into
> something like this:
> 
> (def dat-pivot-desired [[nil :followup-months nil nil]
>                         [:cohort-month "2015-02" "2015-03" "2015-04" ]
>                         ["2015-01" {:x1 105 :x2 207} {:x1 103 :x2 196} {:x1 107 :x2 193}]
>                         ["2015-02" {} {:x1 110 :x2 201} {}]
>                         ])
> 
> Ideally, there could be any number of :x1 ... :xN keys ??
> 
> Thank-you,
> Avram
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cgrand/xforms/issues/2.

## 

On Clojure http://clj-me.cgrand.net/
Clojure Programming http://clojurebook.com
Training, Consulting & Contracting http://lambdanext.eu/

**Comment by aaelony** [comment]
Indeed, nice to re-connect :)

There may be a better desired output, but the motivation is to transform a collection of clojure maps into something Pivot-Table like for inclusion into Tom Faulhaber's Excel Templates project (https://github.com/tomfaulhaber/excel-templates).

Very commonly, the output desired is in the form of a Pivot Table view, hence something like the following for `:x1`:

```
[nil :followup-months nil nil]
[:cohort-month "2015-02" "2015-03" "2015-04" ]
["2015-01" 105 103 107]
["2015-02" nil 110 nil]
```

but on the other hand it's nice to have the "metrics", x1, x2, ... xn, all pre-aggregated, so perhaps nested maps (?).  Perhaps there is a better option than this:

```
[nil :followup-months nil nil]
[:cohort-month "2015-02" "2015-03" "2015-04" ]
["2015-01" {:x1 105 :x2 207} {:x1 103 :x2 196} {:x1 107 :x2 193}]
["2015-02" {} {:x1 110 :x2 201} {}]
```

Also, from there a nice extension would be column percentages and row percentages as well.

Then the pipeline from computation to Excel would be complete (with formatting to boot).

Thank-you for your thoughts.

**Comment by cgrand** [comment]
For a start, this works:

``` clj
=> (into {} (x/by-key (juxt :cohort-month :followup-month)
             #(dissoc % :cohort-month :followup-month :strata)
             (x/reduce (partial merge-with +)))
     dat)
{["2015-01" "2015-02"] {:x1 105, :x2 207}, ["2015-01" "2015-03"] {:x1 103, :x2 196}, ["2015-02" "2015-03"] {:x1 110, :x2 201}, ["2015-01" "2015-04"] {:x1 107, :x2 193}}
```

I'd like to provide a nice way to also compute the sets of cohort and follow-up months in the same pass.

Please note that this partial answer is oblivious to the number of `:xN` keys.

**Comment by cgrand** [comment]
Single pass but way too hacky:

``` clj
(let [just (fn
             ([] nil)
             ([x] x)
             ([_ x] x))]
  (transduce identity (x/juxt-map
                        :data ((comp (x/by-key (juxt :cohort-month :followup-month)
                                       #(dissoc % :cohort-month :followup-month :strata)
                                       (x/reduce (partial merge-with +)))
                                 (x/into {}))
                                just)
                        :cohorts ((comp (map :cohort-month) (x/into #{})) just)
                        :followups ((comp (map :followup-month) (x/into #{})) just)) 
              dat))
```

outputs:

``` clj
{:data {["2015-01" "2015-02"] {:x1 105, :x2 207}, ["2015-01" "2015-03"] {:x1 103, :x2 196}, ["2015-02" "2015-03"] {:x1 110, :x2 201}, ["2015-01" "2015-04"] {:x1 107, :x2 193}}, :cohorts #{"2015-02" "2015-01"}, :followups #{"2015-02" "2015-04" "2015-03"}}
```

**Comment by cgrand** [comment]
I added a new function named `transjuxt`:

``` clj
(x/transjuxt
  {:data (comp (x/by-key (juxt :cohort-month :followup-month)
                 #(dissoc % :cohort-month :followup-month :strata)
                 (x/reduce (partial merge-with +)))
           (x/into {}))
   :cohorts (comp (map :cohort-month) (x/into #{}))
   :followups (comp (map :followup-month) (x/into #{}))} 
  dat)
```

whose output is:

``` clj
{:data {["2015-01" "2015-02"] {:x1 105, :x2 207}, ["2015-01" "2015-03"] {:x1 103, :x2 196}, ["2015-02" "2015-03"] {:x1 110, :x2 201}, ["2015-01" "2015-04"] {:x1 107, :x2 193}},
 :cohorts #{"2015-02" "2015-01"},
 :followups #{"2015-02" "2015-04" "2015-03"}}
```

**Comment by cgrand** [comment]
A full answer would then be:

``` clj
(defn pivot [col row aggregation-xform coll]
  (let [{:keys [cols rows aggs]}
        (x/transjuxt
          {:aggs (comp (x/by-key (juxt col row) #(dissoc % col row)
                         aggregation-xform)
                   (x/into {}))
           :cols (comp (map col) (x/into (sorted-set)))
           :rows (comp (map row) (x/into (sorted-set)))}
          coll)]
    (into [(into [nil col] (repeat (dec (count cols)) nil))
           (into [row] cols)]
      (map (fn [r]
             (into [r] (map #(aggs [% r])) cols)))
      rows)))
```

and:

``` clj
=> (pivot :followup-month :cohort-month
     (comp
       (map #(dissoc % :strata))
       (x/reduce (partial merge-with +)))
     dat)
[[nil :followup-month nil nil]
 [:cohort-month "2015-02" "2015-03" "2015-04"]
 ["2015-01" {:x1 105, :x2 207} {:x1 103, :x2 196} {:x1 107, :x2 193}]
 ["2015-02" nil {:x1 110, :x2 201} nil]]
```

**Comment by aaelony** [comment]
Thank-you, I will study your solution. Looks exactly perfect!!!  Merci!
-Avram

**Comment by aaelony** [comment]
Hi Christophe,

This functionality is fantastic.  I have been noticing that sometimes I get `java.lang.ClassCastException: clojure.lang.PersistentVector cannot be cast to clojure.lang.IPersistentMap` errors though.  

Could it be that transjuxt needs to allow for a lazy sequence as well as a vector?? 

```

(defn transjuxt
  "Performs several transductions over coll at once. xforms-map can be a map or vector.
   Returns a map with the same keyset as xforms-map (or a vector of same size as xforms-map).
   Returns a transducer when coll is omitted."
  ([xforms-map]
    (let [[f args] (if (vector? xforms-map)
                     [juxt (map #(% first))]
                     [juxt-map (comp (by-key (map #(% first))) cat)])]
      (fn [rf]
        ((reduce (apply f (sequence args xforms-map))) rf))))
  ([xforms-map coll]
    (transduce (transjuxt xforms-map) first coll)))

```

Thanks for your thoughts.

Regards,
-Avram

**Comment by cgrand** [comment]
Fixed in 0.1.1 https://github.com/cgrand/xforms/commit/45af15c69db0430f1a73585fdb6e8f62575d6fa7
